pkg/store/lazy_postings.go

```diff
@@ -0,0 +1,272 @@
+
+	   Compare formula F4 (Cost to fetch up to 4 posting groups) and F3.
+	   P4 * 4 < P1 * S * R^2 * (1 - R)
+
+	   We can generalize this to formula: Pn * 4 < P1 * S * R^(n - 2) * (1 - R)
+
+	   The idea of the algorithm:
+	   By iterating the posting group in sorted order of cardinality, we need to make sure that by fetching the current posting group,
+	   the total data fetched is smaller than the previous posting group. If so, then we continue to next posting group,
+	   otherwise we stop.
+
+	   This ensures that when we stop at one posting group, posting groups after it always need to fetch more data.
+	   Based on formula Pn * 4 < P1 * S * R^(n - 2) * (1 - R), left hand side is always increasing while iterating to larger
+	   posting groups while right hand side value is always decreasing as R < 1.
+	*/
+	seriesBytesToFetch := postingGroups[0].cardinality * seriesMaxSize
+	p := float64(1)
+	i := 1 // Start from index 1 as we always need to fetch the smallest posting group.
+	hasAdd := !postingGroups[0].addAll
+	for i < len(postingGroups) {
+		pg := postingGroups[i]
+		// Need to fetch more data on postings than series we avoid fetching, stop here and lazy expanding rest of matchers.
+		// If there is no posting group with add keys, don't skip any posting group until we have one.
+		// Fetch posting group with addAll is much more expensive due to fetch all postings.
+		if hasAdd && pg.cardinality*4 > int64(p*math.Ceil((1-seriesMatchRatio)*float64(seriesBytesToFetch))) {
+			break
+		}
+		hasAdd = hasAdd || !pg.addAll
+		p = p * seriesMatchRatio
+		i++
+	}
+	for i < len(postingGroups) {
+		postingGroups[i].lazy = true
+		lazyExpandedPostingSizeBytes.Add(float64(4 * postingGroups[i].cardinality))
+		i++
+	}
+	return postingGroups, false, nil
+}
+
+func fetchLazyExpandedPostings(
+	ctx context.Context,
+	postingGroups []*postingGroup,
+	r *bucketIndexReader,
+	bytesLimiter BytesLimiter,
+	addAllPostings bool,
+	lazyExpandedPostingEnabled bool,
+	lazyExpandedPostingSizeBytes prometheus.Counter,
+) (*lazyExpandedPostings, error) {
+	var (
+		err               error
+		emptyPostingGroup bool
+	)
+	/*
+			There are several cases that we skip postings fetch optimization:
+			- Lazy expanded posting disabled.
+			- Add all postings. This means we don't have a posting group with any add keys.
+		    - Block estimated max series size not set which means we don't have a way to estimate series bytes downloaded.
+			- `SeriesMaxSize` not set for this block then we have no way to estimate series size.
+			- Only one effective posting group available. We need to at least download postings from 1 posting group so no need to optimize.
+	*/
+	if lazyExpandedPostingEnabled && !addAllPostings &&
+		r.block.estimatedMaxSeriesSize > 0 && len(postingGroups) > 1 {
+		postingGroups, emptyPostingGroup, err = optimizePostingsFetchByDownloadedBytes(
+			r,
+			postingGroups,
+			int64(r.block.estimatedMaxSeriesSize),
+			0.5, // TODO(yeya24): Expose this as a flag.
```

Comment:
Perhaps in the future we could use some kind of heuristics to set this automatically :thinking: